Fix an anomaly caused by command queue

tbg opened the pull request:
Fixes #4607, Fixes #4560.

A recent change the command queue came with an innocuous-looking optimization:
If a new read entered the command queue, it would not wait for previous reads
to overlapping key ranges.

Unfortunately, this introduced the anomaly explained below. Three transactions
A, B and C overlap on a single range, with increasing timestamps. Reads touch
all keys. A only reads; B and C try to read a maximum `m` and then write `m+1`
to their own keys.
The anomaly is that sometimes, B and C can commit with the same new maximum,
which is clearly not linearizable, with the history explained below. The
crux is (***): when B goes in for its write, it doesn't wait for C's read,
which sees the same maximum B read, and should get a chance to update the
timestamp cache. The reason C doesn't wait for B's read is that A went in for
its read after C started, but did not incur a dependency on A (due to the
read-read optimization).

```
+--------------+-----------------------------+
|      A       |       B      |       C      |
---------------+-----------------------------+
|              | enter CQ (r) |              |
|              | read max = 1 |              |
|              | leave CQ (r) |              |
|              |              | enter CQ(r)  |
|              |              | read max = 1 |
| enter CQ (r) |              |              |
|              | enter CQ (w) |              |
|              | wait: A ***  |              |
| read         |              |              |
| leave CQ (r) |              |              |
|              | write 2      |              |
|              | commit       |              |
|              | leave CQ (w) |              |
|              |              | leave CQ (r) |
|              |              | enter CQ (w) |
|              |              | write 2      |
|              |              | commit       |
|              |              | leave CQ (w) |
+--------------+--------------+--------------+
```

The fix is easy: Have write commands wait for all underlying reads, even when
they are already covered.

`Benchmark(Insert|Update|Delete|Scan).*Cockroach` are identical. Ditto for the
Pgbench suite, which had a tiny amount more allocs:

```
name                              old alloc/op   new alloc/op
PgbenchExec_Cockroach-8              152kB ± 0%     153kB ± 0%
PgbenchQuery_Cockroach-8             197kB ± 0%     197kB ± 0%
ParallelPgbenchQuery_Cockroach-8     659kB ± 0%     752kB ± 0%

name                              old allocs/op  new allocs/op
PgbenchExec_Cockroach-8              2.87k ± 0%     2.89k ± 0%
PgbenchQuery_Cockroach-8             3.58k ± 0%     3.57k ± 0%
ParallelPgbenchQuery_Cockroach-8     9.79k ± 0%    10.79k ± 0%
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4651)

<!-- Reviewable:end -->

petermattis commented:
I love that ascii diagram.

nvanbenschoten commented:
LGTM, although you might want to mention that there were two alternatives here, and mention why you chose the second one:
- Remove the read-read independence optimization
- Add dependencies to all overlapping parents for write commands, not just superseding parents

Also, did you happen to get any perf numbers on this alternative and possibly even the other one? If you want to try that out, just remove the readWriteFilter.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/command_queue.go, line 142 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4651#-KBOpO8U9Gnxr9gPwRvn:-KBOpO8V7133rc_mcmTH:655769125) ([raw file](https://github.com/cockroachdb/cockroach/blob/34affedacba70d0051ce35c73093e9dd81d36491/storage/command_queue.go#L142)):*
You'll probably want to put the `!readOnly` first so that this will short circuit correctly.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4651#-:-KBOqP4P9rP6ar1N8K0T:-815110041)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4651#-:-KBOyjZ7qGD-fVCk-iAU:676834233)*
<!-- Sent from Reviewable.io -->

tbg committed:
Fix an anomaly caused by command queue

Fixes #4607, Fixes #4560.

A recent change the command queue came with an innocuous-looking optimization:
If a new read entered the command queue, it would not wait for previous reads
to overlapping key ranges.

Unfortunately, this introduced the anomaly explained below. Three transactions
A, B and C overlap on a single range, with increasing timestamps. Reads touch
all keys. A only reads; B and C try to read a maximum `m` and then write `m+1`
to their own keys.
The anomaly is that sometimes, B and C can commit with the same new maximum,
which is clearly not linearizable, with the history explained below. The
crux is (***): when B goes in for its write, it doesn't wait for C's read,
which sees the same maximum B read, and should get a chance to update the
timestamp cache. The reason C doesn't wait for B's read is that A went in for
its read after C started, but did not incur a dependency on A (due to the
read-read optimization).

```
+--------------+-----------------------------+
|      A       |       B      |       C      |
---------------+-----------------------------+
|              | enter CQ (r) |              |
|              | read max = 1 |              |
|              | leave CQ (r) |              |
|              |              | enter CQ(r)  |
|              |              | read max = 1 |
| enter CQ (r) |              |              |
|              | enter CQ (w) |              |
|              | wait: A ***  |              |
| read         |              |              |
| leave CQ (r) |              |              |
|              | write 2      |              |
|              | commit       |              |
|              | leave CQ (w) |              |
|              |              | leave CQ (r) |
|              |              | enter CQ (w) |
|              |              | write 2      |
|              |              | commit       |
|              |              | leave CQ (w) |
+--------------+--------------+--------------+
```

The fix is easy: Have write commands wait for all underlying reads, even when
they are already covered.

`Benchmark(Insert|Update|Delete|Scan).*Cockroach` are identical. Ditto for the
Pgbench suite, which had a tiny amount more allocs:

```
name                              old alloc/op   new alloc/op
PgbenchExec_Cockroach-8              152kB ± 0%     153kB ± 0%
PgbenchQuery_Cockroach-8             197kB ± 0%     197kB ± 0%
ParallelPgbenchQuery_Cockroach-8     659kB ± 0%     752kB ± 0%

name                              old allocs/op  new allocs/op
PgbenchExec_Cockroach-8              2.87k ± 0%     2.89k ± 0%
PgbenchQuery_Cockroach-8             3.58k ± 0%     3.57k ± 0%
ParallelPgbenchQuery_Cockroach-8     9.79k ± 0%    10.79k ± 0%
```

nvanbenschoten commented:
LGTM

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4651#-:-KBP5w80hPscZJHmxkX1:676834233)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Doesn't this have worst-case performance comparable to the pre-RangeGroup version? Before, every command waited for every other command, which worked out to something like `(r+w)**2`. Now, writes wait for all reads for a total cost of `r*w`. That's fine if either `r` or `w` is small, but if there are equal numbers of reads and writes it would be significant. The optimization mentioned in the TODO comment would address this and so I think it's probably worth doing.

A third, half-baked solution would be to ensure that reads exit the command queue in the same order that they entered. This could be done with a second DAG containing only reads, and commands would wait on this DAG in endCmds instead of beginCmds.

tschottdorf commented:
Another possible venue is adding missing orderings between reads when they are required by a later write. So reads never wait on reads, unless a write tells them to. This should also be possible without larger code changes. Just throwing it out there.
